libs/community/langchain_community/chat_message_histories/sql.py

```diff
@@ -123,18 +184,85 @@ def messages(self) -> List[BaseMessage]:  # type: ignore
                 messages.append(self.converter.from_sql_model(record))
             return messages
 
+    async def aget_messages(self) -> List[BaseMessage]:
+        """Retrieve all messages from db"""
+        assert self.async_mode, "This method must be called with async_mode"
+        await self._acreate_table_if_not_exists()
+        async with self.session_maker() as session:
+            stmt = (
+                select(self.sql_model_class)
+                .where(
+                    getattr(self.sql_model_class, self.session_id_field_name)
+                    == self.session_id
+                )
+                .order_by(self.sql_model_class.id.asc())
+            )
+            result = await session.execute(stmt)
+            messages = []
+            for record in result.scalars():
+                messages.append(self.converter.from_sql_model(record))
+            return messages
+
     def add_message(self, message: BaseMessage) -> None:
         """Append the message to the record in db"""
-        with self.Session() as session:
+        with self.session_maker() as session:
             session.add(self.converter.to_sql_model(message, self.session_id))
             session.commit()
 
+    async def aadd_message(self, message: BaseMessage) -> None:
+        """Add a Message object to the store.
+
+        Args:
+            message: A BaseMessage object to store.
+        """
+        assert self.async_mode, "This method must be called with async_mode"
+        await self._acreate_table_if_not_exists()
+        async with self.session_maker() as session:
+            session.add(self.converter.to_sql_model(message, self.session_id))
+            await session.commit()
+
+    def add_messages(self, messages: Sequence[BaseMessage]) -> None:
+        # The method RunnableWithMessageHistory._exit_history() call
+        #  add_message method by mistake and not aadd_message.
+        # See https://github.com/langchain-ai/langchain/issues/22021
+        if self.async_mode:
+            loop = asyncio.get_event_loop()
```

Comment:
We generally raise exceptions here instead of kicking off another event loop.

We need to resolve a design issue with `RunnableWithMessageHistory` before I'd feel OK breaking our rule of generally not invoking async code from sync code

Reply:
Ok, but you must fix the bug describe [here](https://github.com/langchain-ai/langchain/issues/22021), in `_exit_history()`.
I'd be delighted to remove this ugly code ;-)

but for the moment it's an acceptable temporary solution.

Reply:
ok that's reasonable